metabase/metabase
# [MLv2] Filters — Re-use values when switching filter operators

Fixes #34959

We used to wipe the filter pickers' `values` state when changing an operator. An input component would receive an `undefined` value, become uncontrolled, and enter an unexpected UI state (input shows the previous value, but the "Update filter" button is disabled).

The PR updates filter picker components, so they do their best to re-use `values` when an operator changes. Also, it updates the filter picker's `constants/OPERATOR_OPTION` to be a map instead of an array. I think the map turned out to be a handier format for everything except the `BooleanFilterPicker`.

### To Verify

1. New > Question > Raw Data > Sample Database > Orders
2. Add a filter for a numeric column, e.g. Subtotal > 10
3. Change the operator to something like "Equals to"
4. Ensure the value 10 is re-used, and you can still update the filter
5. Repeat for a string, time, and coordinate columns with different operators

### Demo

https://github.com/metabase/metabase/assets/17258145/3f948ec9-bf72-4cbb-920e-2d253e135ac8
